### PR TITLE
Add teletext and teletext-yle recipes

### PR DESCRIPTION
**Pull request by lassik**
### Brief summary of what the package does

Lets people interactively browse [teletext](https://en.wikipedia.org/wiki/Teletext) pages in Emacs.

### Direct link to the package repository

https://github.com/lassik/emacs-teletext -- the teletext framework

https://github.com/lassik/emacs-teletext-yle -- a plug-in to provide pages from the Finnish national TV network

(More plug-ins could be written as separate Emacs packages to cover more TV networks.)

### Your association with the package

Author

### Relevant communications with the upstream package maintainer

None needed

### Checklist

Please confirm with `x`:

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses).
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] `M-x checkdoc` is happy with my docstrings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [ ] I have confirmed some of these without doing them

### CODE REVIEW

**Comment by lassik** [comment]
**NOTE:** There's currently a problem with [this part](https://github.com/lassik/emacs-teletext-yle/blob/5725459279747a5385b22ace2c045f5d4dabaa3c/teletext-yle.el#L191) of the plug-in. The idea is:

* The `teletext-yle` plug-in package requires the main `teletext` package.
* When `teletext-yle` is loaded, it registers itself by calling the `teletext-provide` function from the main `teletext` package.
* In the main `teletext` package, the `teletext` command is autoloaded. This means that `M-x teletext` loads the package. If `teletext-yle` is installed well, the main package's autoload should trigger the loading of `teletext-yle` as well.

However, I can't figure out how to get the autoloading to work. I tried `eval-when-compile` and `eval-after-load`; neither works. What's the usual way to do this?

**Comment by riscy** [comment]
> the the main package's autoload should trigger the loading of teletext-yle as well.

I'm not sure I fully understand, but let me rephrase to see if I've got it: a user loads in teletext-yle.  teletext-yle loads teletext through the `require` on it (regardless of autoload cookies).  teletext-yle registers itself with teletext.

I think everything should work out but I'm probably missing something -- which aspect of autoloading isn't working?

**Comment by lassik** [comment]
From a user's point of view, if they have installed both the `teletext` and `teletext-yle` packages, then doing `M-x teletext` (which autoloads the `teletext` package) would ideally cause the `teletext-yle` package to be loaded as well.

I looked at popular packages like `company` that have add-on packages like this, but couldn't find one where the add-ons are loaded automatically. Is it customary that the add-ons need to be explicitly `require`d in user init files?

**Comment by riscy** [comment]
> Is it customary that the add-ons need to be explicitly required in user init files?

Yeah, this is the usual case -- it's up to the users or starter-pack authors to do this.  This pushes a bit of complexity back onto them, but keeps the control in their hands and makes it easier for them to debug.

I'd actually discourage breaking with this convention.  I'm sure some packages do, and I'm sure some of those packages are also the reason some Emacs newcomers can end up in confusing debugging situations.  But if you _really_ wanted to, you can do things like write the registry to a file (or define the list of registered plugins as a customizable with `defcustom`, and then write to the user's config with `customize-save-variable`).  Or you could look at the code for `load-library` and have teletext scan for plugins on the `load-path`.

**Comment by lassik** [comment]
> > Is it customary that the add-ons need to be explicitly required in user init files?
> 
> Yeah, this is the usual case -- it's up to the users or starter-pack authors to do this. This pushes a bit of complexity back onto them, but keeps the control in their hands and makes it easier for them to debug.

Sounds reasonable.

> I'd actually discourage breaking with this convention. I'm sure some packages do, and I'm sure some of those packages are also the reason some Emacs newcomers can end up in confusing debugging situations. But if you _really_ wanted to, you can do things like write the registry to a file (or define the list of registered plugins as a customizable with `defcustom`, and then write to the user's config with `customize-save-variable`). Or you could look at the code for `load-library` and have teletext scan for plugins on the `load-path`.

Nah, I now strongly agree with your judgement that it's simpler to have users manually load the add-on packages :) Thanks for making your case in detail. It is convincing.

What should `teletext-yle` do specifically:

* Is it enough that users `(require 'teletext-yle)` in their init file, and `teletext-yle.el` would eval some code at load time to register itself?

* Or should `teletext-yle` have a public function `teletext-yle-provide` (or similar) which users can call in their init file to load the add-on? If so, should the function have autoload?

**Comment by riscy** [comment]
Thanks for being accommodating. :)  I'd recommend the latter -- usually merely loading a file shouldn't modify Emacs's behavior (again for debuggability).  There are exceptions to this rule, like color themes registering themselves, and auto-mode-alist.  But doing so explicitly in a `-provide`/`-unprovide` like you describe, or in some cases a minor mode that can be toggled off so the behavior can be undone, is the safe way to go.

**Comment by lassik** [comment]
Good advice. Let's do that. Should the function be autoloaded, and do people still need to `(require 'teletext-yle)` in their init file in addition to calling the function?

**Comment by riscy** [comment]
> Should the function be autoloaded

Yep, this makes sense.

> and do people still need to (require 'teletext-yle) in their init file in addition to calling the function?

Nope, assuming they've formally installed it and an [autoloads.el file has been written](https://www.gnu.org/software/emacs/manual/html_node/elisp/Autoload.html), they don't need to.

**Comment by lassik** [comment]
OK, done!

**Comment by riscy** [comment]
Thanks.  This looks good to me.
